File: rust_tm_api/benchmarks/src/stamp/bayes.rs
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::Arc;
use std::thread_local;
use std::cell::RefCell;
use tm::{transaction, TmCell};
use crate::Rng;
use super::Config;

const MAX_TASKS: usize = 131072;
const MAX_PARENTS: usize = 2;

// ── Small fixed-size parent set (TM-tracked) ──────────────
struct ParentSet {
    count: TmCell<i32>,
    data: [TmCell<i32>; MAX_PARENTS],
}

impl ParentSet {
    fn new() -> Self {
        ParentSet { count: TmCell::new(0), data: [TmCell::new(-1), TmCell::new(-1)] }
    }
    fn contains(&self, tx: &tm::Transaction, val: i32) -> bool {
        let n = tx.read(&self.count);
        for i in 0..n as usize {
            if tx.read(&self.data[i]) == val { return true; }
        }
        false
    }
    fn insert(&self, tx: &tm::Transaction, val: i32) {
        let n = tx.read(&self.count) as usize;
        if n < MAX_PARENTS {
            tx.write(&self.data[n], val);
            tx.write(&self.count, (n + 1) as i32);
        }
    }
    fn peek_count(&self) -> i32 { unsafe { *self.count.ptr() } }
    fn peek_get(&self, i: usize) -> i32 { unsafe { *self.data[i].ptr() } }
    fn peek_insert(&self, val: i32) {
        let n = self.peek_count() as usize;
        if n < MAX_PARENTS {
            unsafe { *self.data[n].ptr() = val; }
            unsafe { *self.count.ptr() = (n + 1) as i32; }
        }
    }
    fn collect(&self, tx: &tm::Transaction) -> Vec<i32> {
        let n = tx.read(&self.count) as usize;
        (0..n).map(|i| tx.read(&self.data[i])).collect()
    }
}

// ── Task list (TM priority queue) ─────────────────────────
struct TaskList {
    scores: Vec<TmCell<f64>>,
    ops: Vec<TmCell<i32>>,
    from_ids: Vec<TmCell<i32>>,
    to_ids: Vec<TmCell<i32>>,
    count: TmCell<i32>,
}

#[derive(Clone)]
struct Task { op: i32, from: i32, to: i32, score: f64 }

impl TaskList {
    fn new() -> Self {
        let n = MAX_TASKS;
        let mut scores = Vec::with_capacity(n);
        let mut ops = Vec::with_capacity(n);
        let mut from_ids = Vec::with_capacity(n);
        let mut to_ids = Vec::with_capacity(n);
        for _ in 0..n {
            scores.push(TmCell::new(0.0));
            ops.push(TmCell::new(0));
            from_ids.push(TmCell::new(0));
            to_ids.push(TmCell::new(0));
        }
        TaskList { scores, ops, from_ids, to_ids, count: TmCell::new(0) }
    }
    fn push(&self, tx: &tm::Transaction, op: i32, from: i32, to: i32, score: f64) {
        let i = tx.read(&self.count) as usize;
        if i >= MAX_TASKS { return; }
        tx.write(&self.scores[i], score);
        tx.write(&self.ops[i], op);
        tx.write(&self.from_ids[i], from);
        tx.write(&self.to_ids[i], to);
        tx.write(&self.count, (i + 1) as i32);
    }
    fn push_setup(&self, op: i32, from: i32, to: i32, score: f64) {
        let i = unsafe { *self.count.ptr() } as usize;
        if i >= MAX_TASKS { return; }
        unsafe { *self.scores[i].ptr() = score; }
        unsafe { *self.ops[i].ptr() = op; }
        unsafe { *self.from_ids[i].ptr() = from; }
        unsafe { *self.to_ids[i].ptr() = to; }
        unsafe { *self.count.ptr() = (i + 1) as i32; }
    }
    fn pop_best(&self, tx: &tm::Transaction) -> Task {
        let n = tx.read(&self.count) as usize;
        if n == 0 { return Task { op: -1, from: -1, to: -1, score: -1e100 }; }
        let mut best = 0usize;
        let mut best_score = tx.read(&self.scores[0]);
        for i in 1..n {
            let s = tx.read(&self.scores[i]);
            if s > best_score { best_score = s; best = i; }
        }
        let last = n - 1;
        let t = Task {
            op: tx.read(&self.ops[best]),
            from: tx.read(&self.from_ids[best]),
            to: tx.read(&self.to_ids[best]),
            score: best_score,
        };
        if best != last {
            tx.write(&self.scores[best], tx.read(&self.scores[last]));
            tx.write(&self.ops[best], tx.read(&self.ops[last]));
            tx.write(&self.from_ids[best], tx.read(&self.from_ids[last]));
            tx.write(&self.to_ids[best], tx.read(&self.to_ids[last]));
        }
        tx.write(&self.count, last as i32);
        t
    }
    fn peek_empty(&self) -> bool { let v = unsafe { *self.count.ptr() }; v == 0 }
}

// ── Bayes data ────────────────────────────────────────────
struct BayesData {
    num_var: usize,
    num_records: usize,
    base_penalty: f64,
    records: Vec<Vec<i32>>,
    parents: Vec<ParentSet>,
    children: Vec<ParentSet>,
    local_ll: Vec<TmCell<f64>>,
    base_log_likelihood: TmCell<f64>,
    total_parents: TmCell<i32>,
    global_max_edges: i32,
    task_list: TaskList,
}

// ── Compute density log-likelihood for a variable with given parents ─
fn compute_density_ll(data: &BayesData, var: usize, parents: &[i32]) -> f64 {
    let num_configs = 1usize << parents.len();
    let mut cfg = vec![(0i32, 0i32); num_configs];
    for r in 0..data.num_records {
        let mut config = 0usize;
        for (i, &p) in parents.iter().enumerate() {
            if data.records[r][p as usize] != 0 { config |= 1 << i; }
        }
        if data.records[r][var] != 0 {
            cfg[config].1 += 1;
        } else {
            cfg[config].0 += 1;
        }
    }
    let mut ll = 0.0f64;
    let nf = data.num_records as f64;
    for c in 0..num_configs {
        let total = cfg[c].0 + cfg[c].1;
        if total == 0 { continue; }
        let frac = total as f64 / nf;
        let p0 = cfg[c].0 as f64 / total as f64;
        let p1 = cfg[c].1 as f64 / total as f64;
        if p0 > 0.0 { ll += frac * p0 * p0.ln(); }
        if p1 > 0.0 { ll += frac * p1 * p1.ln(); }
    }
    ll
}

// ── Cycle detection (DFS, reads committed child sets) ─────
fn has_path(data: &BayesData, from: i32, to: i32) -> bool {
    thread_local! {
        static VISITED: RefCell<Vec<bool>> = RefCell::new(Vec::new());
        static STACK: RefCell<Vec<i32>> = RefCell::new(Vec::new());
    }
    VISITED.with(|v| {
        let mut visited = v.borrow_mut();
        visited.clear();
        visited.resize(data.num_var, false);
        STACK.with(|s| {
            let mut stack = s.borrow_mut();
            stack.clear();
            stack.push(from);
            while let Some(cur) = stack.pop() {
                if cur == to { return true; }
                if cur as usize >= data.num_var { continue; }
                if visited[cur as usize] { continue; }
                visited[cur as usize] = true;
                let n = data.children[cur as usize].peek_count();
                for i in 0..n as usize {
                    let child = data.children[cur as usize].peek_get(i);
                    if child >= 0 && !visited[child as usize] {
                        stack.push(child);
                    }
                }
            }
            false
        })
    })
}

// ── Find best insertion for a variable ────────────────────
fn find_best_insert(data: &BayesData, to: i32, tx: &tm::Transaction) -> Task {
    if tx.read(&data.total_parents) >= data.global_max_edges * data.num_var as i32 {
        return Task { op: -1, from: -1, to: -1, score: -1e100 };
    }
    let mut best = Task { op: 0, from: -1, to: -1, score: -1e100 };
    for from in 0..data.num_var as i32 {
        if from == to { continue; }
        if data.parents[to as usize].contains(tx, from) { continue; }
        if has_path(data, to, from) { continue; }
        let mut par = data.parents[to as usize].collect(tx);
        par.push(from);
        let new_ll = compute_density_ll(data, to as usize, &par);
        let delta = new_ll - tx.read(&data.local_ll[to as usize]);
        let score = tx.read(&data.total_parents) as f64 * data.base_penalty
                  + data.num_records as f64 * (tx.read(&data.base_log_likelihood) + delta);
        if score > best.score {
            best = Task { op: 0, from, to, score };
        }
    }
    best
}

// ── Worker function ───────────────────────────────────────
pub fn run(config: &Config, stop: &AtomicBool, ops: &AtomicU64) {
    println!("\n=== Bayes ===");
    let num_var = config.points.max(8).min(64);
    let num_records = config.num_customers.max(64).min(2048);
    let max_parents = 2usize;
    let insert_penalty = 2.0;
    let base_penalty = -0.5 * (num_records as f64).ln() * insert_penalty;
    let global_max_edges = 2i32;
    println!("  Vars: {}  Records: {}  Max parents: {}  Penalty: {:.0}",
             num_var, num_records, max_parents, insert_penalty);

    // ── Data generation ──────────────────────────────────
    let mut rng = Rng::new(42);
    // Generate random parent graph
    let mut parents_init: Vec<Vec<i32>> = (0..num_var).map(|_| Vec::new()).collect();
    for v in 1..num_var {
        let n = (rng.next() as usize % max_parents.min(v)) + 1;
        for _ in 0..n {
            loop {
                let p = (rng.next() as usize) % v;
                if !parents_init[v].contains(&(p as i32)) {
                    parents_init[v].push(p as i32);
                    break;
                }
            }
        }
    }
    // Generate random records
    let records: Vec<Vec<i32>> = (0..num_records).map(|_| {
        (0..num_var).map(|v| {
            if parents_init[v].is_empty() {
                (rng.next() % 2) as i32
            } else {
                let mut key = rng.next();
                for &p in &parents_init[v] {
                    key = key.wrapping_mul(31).wrapping_add(key);
                }
                (key % 2) as i32
            }
        }).collect()
    }).collect();

    // ── Allocate TM data structures ──────────────────────
    let parents: Vec<ParentSet> = (0..num_var).map(|_| ParentSet::new()).collect();
    let children: Vec<ParentSet> = (0..num_var).map(|_| ParentSet::new()).collect();
    let local_ll: Vec<TmCell<f64>> = (0..num_var).map(|_| TmCell::new(0.0)).collect();
    let base_log_likelihood = TmCell::new(0.0);
    let total_parents = TmCell::new(0);
    let task_list = TaskList::new();
    let data = Arc::new(BayesData {
        num_var, num_records, base_penalty,
        records: records.clone(),
        parents, children, local_ll,
        base_log_likelihood, total_parents,
        global_max_edges, task_list,
    });

    // Initialize parent sets from generated graph (non-TM)
    for v in 0..num_var {
        for &p in &parents_init[v] {
            data.parents[v].peek_insert(p);
            data.children[p as usize].peek_insert(v as i32);
        }
    }

    let dur = std::time::Duration::from_millis(config.duration as u64);
    let g_ops = AtomicU64::new(0);
    let t0 = std::time::Instant::now();

    std::thread::scope(|s| {
        let mut threads = Vec::new();
        for tid in 0..config.threads.max(1).min(2) { // at most 2 threads for bayes
            let d = data.clone();
            let sc = stop;
            let so = ops.clone();
            let go = &g_ops;
            let chunk = num_var / config.threads.max(1);
            let start = tid * chunk;
            let end = if tid == config.threads - 1 { num_var } else { (start + chunk).min(num_var) };
            threads.push(s.spawn(move || {
                if tid == 0 {
                    // Phase 1: Initial per-variable LL (just thread 0)
                    // Already initialized in setup, nothing to do here.
                }

                // Phase 2: Enqueue single-edge tasks (parallel)
                if tid == 0 && end > start {
                    // Phase 2 + Phase 3 happen in thread 0
                    thread2(&d, start, end, go, sc, so);
                }
            }));
        }
        std::thread::sleep(dur);
        stop.store(true, Ordering::Relaxed);
    });
    let elapsed = t0.elapsed().as_millis() as u64;
    let ops_count = g_ops.load(Ordering::Relaxed);
    let total_par = unsafe { *data.total_parents.ptr() };
    println!("  Operations: {}  Total parents: {}  Elapsed: {} ms",
             ops_count, total_par, elapsed);
}

fn thread2(data: &BayesData, _start: usize, _end: usize,
           g_ops: &AtomicU64, stop: &AtomicBool, _ops: &AtomicU64) {
    // Initialize per-variable LL
    let mut init_rng = Rng::new(42);
    let _ = &mut init_rng;
    let mut base_ll = 0.0f64;
    for v in 0..data.num_var {
        let ll = compute_density_ll(data, v, &[]);
        transaction(|tx| {
            tx.write(&data.local_ll[v], ll);
            tx.write(&data.base_log_likelihood, tx.read(&data.base_log_likelihood) + ll);
        });
        base_ll += ll;
    }
    // Enqueue single-edge tasks (Phase 2)
    for v in 0..data.num_var {
        for from in 0..data.num_var as i32 {
            if from == v as i32 { continue; }
            let with_ll = compute_density_ll(data, v, &[from]);
            if with_ll > base_ll / data.num_var as f64 {
                let delta = with_ll - base_ll / data.num_var as f64;
                let score = data.base_penalty + data.num_records as f64 * (base_ll + delta);
                transaction(|tx| {
                    data.task_list.push(tx, 0, from, v as i32, score);
                });
            }
        }
    }
    // Phase 3: Greedy search loop
    while !stop.load(Ordering::Relaxed) {
        // Pop best task
        let task = transaction(|tx| data.task_list.pop_best(tx));
        if task.op < 0 { break; }

        // Validate and apply
        transaction(|tx| {
            if data.parents[task.to as usize].contains(tx, task.from) { return; }
            if has_path(data, task.to, task.from) { return; }
            if tx.read(&data.total_parents) >= data.global_max_edges * data.num_var as i32 { return; }

            data.parents[task.to as usize].insert(tx, task.from);
            data.children[task.from as usize].insert(tx, task.to);
            tx.write(&data.total_parents, tx.read(&data.total_parents) + 1);

            // Recompute LL with new parent set
            let par = data.parents[task.to as usize].collect(tx);
            let new_ll = compute_density_ll(data, task.to as usize, &par);
            let delta = new_ll - tx.read(&data.local_ll[task.to as usize]);
            tx.write(&data.local_ll[task.to as usize], new_ll);
            tx.write(&data.base_log_likelihood, tx.read(&data.base_log_likelihood) + delta);

            // Find next best for same variable
            let next = find_best_insert(data, task.to, tx);
            if next.from >= 0 {
                data.task_list.push(tx, next.op, next.from, next.to, next.score);
            }
        });
        g_ops.fetch_add(1, Ordering::Relaxed);
    }
}

File: expli-benchmarks/STAMP/bayes/bayes.cpp
#include "expli_tm_api/tm_api.hpp"
#include "expli_tm_api/containers/tm_small_set.hpp"
#include "expli_tm_api/thread_barrier.hpp"

#include <algorithm>
#include <atomic>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <random>
#include <thread>
#include <vector>

template <typename F>
inline void tx_retry(F&& body) {
    tm_nested_call_counter++;
    int done = 0;
    while (!done) {
        tm_longjmp_ret = sigsetjmp(tm_jmpbuf, 0);
        tm_begin();
        if (tm_longjmp_ret != 0)
            continue;
        body();
        tm_end();
        done = 1;
    }
    tm_nested_call_counter--;
}

struct Config {
    int threads = 4;
    int vars = 32;
    int records = 1024;
    int max_parents = 2;
    int insert_penalty = 2;
    int max_edges_per_var = 2;
    unsigned seed = 0;
};

static Config parse_args(int argc, char *argv[]) {
    Config c;
    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "-t") && i+1 < argc) c.threads = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-v") && i+1 < argc) c.vars = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-r") && i+1 < argc) c.records = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-n") && i+1 < argc) c.max_parents = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-p") && i+1 < argc) c.insert_penalty = atoi(argv[++i]);
        else if (!strcmp(argv[i], "-s") && i+1 < argc) c.seed = (unsigned)atoi(argv[++i]);
    }
    return c;
}

struct Task {
    int op;
    int from_id;
    int to_id;
    double score;
};

// TM-friendly task list (replaces std::vector + std::mutex).
struct TaskList {
    static constexpr int MAX = 131072;
    expli::TM<double> scores[MAX];
    expli::TM<int> ops[MAX];
    expli::TM<int> from_ids[MAX];
    expli::TM<int> to_ids[MAX];
    expli::TM<int> count;

    TaskList() : count() { count.poke(0); }

    bool empty() { return count.read() == 0; }

    void push(int op, int from, int to, double score) {
        int i = count.read();
        if (i >= MAX) return;
        scores[i].write(score);
        ops[i].write(op);
        from_ids[i].write(from);
        to_ids[i].write(to);
        count.write(i + 1);
    }

    // Non-TM setup push (call only outside TX, before threads start).
    void setup_push(int op, int from, int to, double score) {
        int i = count.peek();
        if (i >= MAX) return;
        scores[i].poke(score);
        ops[i].poke(op);
        from_ids[i].poke(from);
        to_ids[i].poke(to);
        count.poke(i + 1);
    }

    Task pop_best() {
        int n = count.read();
        if (n == 0) return {-1, -1, -1, -1e100};
        int best = 0;
        double best_score = scores[0].read();
        for (int i = 1; i < n; i++) {
            double s = scores[i].read();
            if (s > best_score) { best_score = s; best = i; }
        }
        Task t = {ops[best].read(), from_ids[best].read(), to_ids[best].read(), best_score};
        int last = n - 1;
        if (best != last) {
            scores[best].write(scores[last].read());
            ops[best].write(ops[last].read());
            from_ids[best].write(from_ids[last].read());
            to_ids[best].write(to_ids[last].read());
        }
        count.write(last);
        return t;
    }
};

struct BayesData {
    int num_var;
    int num_record;
    double base_penalty;

    std::vector<std::vector<int>> records;

    std::vector<TMSmallSet<int, 2>> parents;
    std::vector<TMSmallSet<int, 2>> children;
    std::vector<expli::TM<double>> local_ll;
    expli::TM<double> base_log_likelihood;
    expli::TM<int> total_parents;

    int global_max_edges;
    TaskList task_list;
};

static BayesData *g_data = nullptr;

static double compute_density_ll(BayesData* data, int var,
                                  const std::vector<int>& parents_vec) {
    int np = (int)parents_vec.size();
    int ncfg = 1 << np;
    std::vector<int> c0(ncfg, 0), c1(ncfg, 0);

    for (int r = 0; r < data->num_record; r++) {
        int cfg = 0;
        for (int i = 0; i < np; i++)
            cfg = (cfg << 1) | data->records[r][parents_vec[i]];
        if (data->records[r][var] == 0) c0[cfg]++; else c1[cfg]++;
    }

    double ll = 0.0;
    for (int c = 0; c < ncfg; c++) {
        int t = c0[c] + c1[c];
        if (t == 0) continue;
        double p0 = (double)c0[c] / t;
        double p1 = (double)c1[c] / t;
        double frac = (double)t / data->num_record;
        if (p0 > 0) ll += frac * p0 * std::log(p0);
        if (p1 > 0) ll += frac * p1 * std::log(p1);
    }
    return ll;
}

static bool has_path(BayesData* data, int from, int to) {
    thread_local std::vector<bool> visited;
    visited.assign(data->num_var, false);
    thread_local std::vector<int> stack;
    stack.clear();
    stack.push_back(from);
    visited[from] = true;
    while (!stack.empty()) {
        int cur = stack.back();
        stack.pop_back();
        int nc = data->children[cur].count();
        for (int i = 0; i < nc; i++) {
            int c = data->children[cur].get(i);
            if (c == to) return true;
            if (!visited[c]) { visited[c] = true; stack.push_back(c); }
        }
    }
    return false;
}

static Task find_best_insert(BayesData* data, int to) {
    Task best = {-1, -1, -1, -1e100};
    double base_ll = data->local_ll[to].read();
    int np = data->parents[to].count();

    if (data->global_max_edges > 0 &&
        data->parents[to].count() >= data->global_max_edges)
        return best;

    for (int from = 0; from < data->num_var; from++) {
        if (from == to) continue;
        if (data->parents[to].contains(from)) continue;
        if (has_path(data, to, from)) continue;

        std::vector<int> par;
        data->parents[to].collect(par);
        par.push_back(from);
        double new_ll = compute_density_ll(data, to, par);
        double delta = new_ll - base_ll;
        double score = data->total_parents.read() * data->base_penalty +
                       data->num_record * (data->base_log_likelihood.read() + delta);

        if (score > best.score)
            best = {0, from, to, score};
    }
    return best;
}

static std::atomic<uint64_t> g_ops{0};

static void worker(int tid, const Config &cfg, expli::Barrier *barrier) {
    tm_init_thread();

    auto data = g_data;
    int nvar = data->num_var;
    int chunk = (nvar + cfg.threads - 1) / cfg.threads;
    int start = tid * chunk;
    int end = std::min(start + chunk, nvar);

    for (int v = start; v < end; v++) {
        double ll = compute_density_ll(data, v, {});
        tx_retry([&]() {
            data->local_ll[v].write(ll);
            data->base_log_likelihood.write(
                data->base_log_likelihood.read() + ll);
        });
    }

    barrier->wait();

    for (int v = start; v < end; v++) {
        for (int from = 0; from < nvar; from++) {
            if (from == v) continue;
            double with_ll = compute_density_ll(data, v, {from});
            double base_ll;
            tx_retry([&]() {
                base_ll = data->local_ll[v].read();
            });
            if (with_ll > base_ll) {
                double delta = with_ll - base_ll;
                double score;
                tx_retry([&]() {
                    score = data->base_penalty +
                            data->num_record * (data->base_log_likelihood.read() + delta);
                });
                tx_retry([&]() {
                    data->task_list.push(0, from, v, score);
                });
            }
        }
    }

    barrier->wait();

    if (tid == 0) {
        for (;;) {
            Task t;
            bool have_task = false;
            tx_retry([&]() {
                if (data->task_list.empty()) return;
                t = data->task_list.pop_best();
                have_task = true;
            });
            if (!have_task) break;

            bool ok = false;
            tx_retry([&]() {
                if (data->parents[t.to_id].contains(t.from_id)) return;
                if (has_path(data, t.to_id, t.from_id)) return;
                if (data->global_max_edges > 0 &&
                    data->parents[t.to_id].count() >= data->global_max_edges) return;

                data->parents[t.to_id].insert(t.from_id);
                data->children[t.from_id].insert(t.to_id);
                data->total_parents.write(data->total_parents.read() + 1);

                std::vector<int> par;
                data->parents[t.to_id].collect(par);
                double new_ll = compute_density_ll(data, t.to_id, par);
                data->base_log_likelihood.write(
                    data->base_log_likelihood.read() +
                    (new_ll - data->local_ll[t.to_id].read()));
                data->local_ll[t.to_id].write(new_ll);

                ok = true;
            });

            if (ok) {
                Task next;
                bool has_next = false;
                tx_retry([&]() {
                    next = find_best_insert(data, t.to_id);
                    has_next = next.op >= 0;
                });
                if (has_next) {
                    tx_retry([&]() {
                        data->task_list.push(next.op, next.from_id, next.to_id, next.score);
                    });
                }
                g_ops.fetch_add(1, std::memory_order_relaxed);
            }
        }
    }
    tm_exit_thread();
}

int main(int argc, char *argv[]) {
    auto cfg = parse_args(argc, argv);

    printf("Bayes — Explicit TM API (TM-friendly data structures)\n");
    printf("Threads: %d  Vars: %d  Records: %d  MaxParents: %d  Penalty: %d\n",
           cfg.threads, cfg.vars, cfg.records, cfg.max_parents, cfg.insert_penalty);

    auto data = new BayesData();
    data->num_var = cfg.vars;
    data->num_record = cfg.records;
    data->base_penalty = cfg.insert_penalty > 0
        ? -0.5 * std::log((double)cfg.records) * cfg.insert_penalty
        : 0.0;
    data->base_log_likelihood.poke(0.0);
    data->total_parents.poke(0);
    data->global_max_edges = cfg.max_edges_per_var;

    data->parents.resize(data->num_var, TMSmallSet<int, 2>());
    data->children.resize(data->num_var, TMSmallSet<int, 2>());
    data->local_ll.resize(data->num_var);
    for (auto &ll : data->local_ll) ll.poke(0.0);

    auto rng = std::mt19937(cfg.seed);
    data->records.resize(data->num_record, std::vector<int>(data->num_var));

    // ── Generate random parent graph (non-TM setup) ──────────
    for (int v = 1; v < data->num_var; v++) {
        int max_np = std::min(cfg.max_parents, v);
        int np = (max_np > 0) ? (int)(rng() % max_np) + 1 : 0;
        for (int p = 0; p < np && p < v; p++) {
            int parent = (int)(rng() % v);
            if (data->parents[v].setup_contains(parent)) continue;
            data->parents[v].setup_insert(parent);
            data->children[parent].setup_insert(v);
            data->total_parents.poke(data->total_parents.peek() + 1);
        }
    }

    // ── Generate random records (non-TM) ─────────────────────
    for (int r = 0; r < data->num_record; r++) {
        for (int v = 0; v < data->num_var; v++) {
            if (data->parents[v].setup_count() == 0) {
                data->records[r][v] = (int)(rng() % 2);
            } else {
                int np = data->parents[v].setup_count();
                int val = 0;
                for (int p = 0; p < np; p++) {
                    int pid = data->parents[v].setup_get(p);
                    val = (val << 1) | data->records[r][pid];
                }
                int threshold = 30;
                for (int p = 0; p < np; p++) {
                    int pid = data->parents[v].setup_get(p);
                    threshold += (pid * 7 + v * 11) % 40;
                }
                threshold = (threshold + val * 17) % 100;
                data->records[r][v] = (rng() % 100 < threshold) ? 1 : 0;
            }
        }
    }

    g_data = data;

    // ── Compute initial local_ll and fill task list (non-TM) ─
    for (int v = 0; v < data->num_var; v++) {
        double ll0 = compute_density_ll(data, v, {});
        data->local_ll[v].poke(ll0);
        data->base_log_likelihood.poke(
            data->base_log_likelihood.peek() + ll0);
    }

    for (int v = 0; v < data->num_var; v++) {
        for (int from = 0; from < data->num_var; from++) {
            if (from == v) continue;
            double with_ll = compute_density_ll(data, v, {from});
            if (with_ll > data->local_ll[v].peek()) {
                double delta = with_ll - data->local_ll[v].peek();
                double score = data->base_penalty +
                               data->num_record * (data->base_log_likelihood.peek() + delta);
                data->task_list.setup_push(0, from, v, score);
            }
        }
    }

    printf("Generating data... done (%d initial tasks).\n",
           data->task_list.count.peek());
    printf("Learning structure...\n");
    fflush(stdout);

    tm_init();
    expli::Barrier barrier(cfg.threads);

    auto start = std::chrono::high_resolution_clock::now();
    std::vector<std::thread> threads;
    for (int t = 0; t < cfg.threads; t++)
        threads.emplace_back(worker, t, std::ref(cfg), &barrier);
    for (auto &th : threads)
        th.join();
    tm_exit();
    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();

    uint64_t ops = g_ops.load();
    int total_parents = data->total_parents.peek();
    printf("\nResults (%lld ms):\n", (long long)elapsed);
    printf("  Operations: %llu  Total parents: %d\n",
           (unsigned long long)ops, total_parents);
    printf("  PASS\n");

    delete data;
    return 0;
}
